keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1169 @@
+
+##### Write Deadline
+
+`SetWriteDeadline` sets up a `time.AfterFunc` [1] with the given deadline, once
+the write deadline exceeds the stream is closed [2] with a `stream reset` error,
+and the client sees the stream error immediately, although the request handler
+on the server might still be running after the timeout happens.
+
+If the request handler on the server continues to write to the underlying
+`ResponseWriter` object, the `Write` method will eventually return an
+`i/o timeout` error (as soon as the internal buffer of the `ResponseWriter` 
+object is full).
+
+The `FlushError` method, on the other hand, is expected to return
+an `i/o timeout` error immediately.
+
+[1] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/h2_bundle.go#L6570-L6594
+
+[2] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/h2_bundle.go#L5691-L5699
+
+
+##### Read Deadline
+
+`SetReadDeadline` sets up a `time.AfterFunc` [1] with the given deadline, once 
+the read deadline exceeds, it closes the request `Body` with
+an `i/o timeout` error [2].
+
+> Note that in the case of read deadline exceeding, the stream is not closed with a `stream reset` error
+
+
+[1] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/h2_bundle.go#L6544-L6568
+
+[2] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/h2_bundle.go#L5681-L5689
+
+
+##### Integrate FlushError
+
+The `Flush` method of the underlying `ResponseWriter` object does not return an
+`error`. With the introduction of per-request read/write deadline the golang team
+has added a new version of `Flush` that returns an `error`, for both http/1x and http/2.0
+
+```
+FlushError() error
+```
+
+[1] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/server.go#L1719C20-L1729
+
+[2] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/h2_bundle.go#L6600-L6623
+
+
+We need to replace the use of `Flush` with `FlushError`:
+```
+// Flush
+flusher, ok := w.(http.Flusher)
+flusher.Flush()
+
+// Replace with FlushError
+flusher, ok := w.(interface{ FlushError() error })
+if err := flusher.FlushError(); err != nil {
+	return err
+}
+```
+
+> With the use of `FlushError`, the request handler can terminate as soon as it 
+> receives an error from `FlushError` after the timeout happens.
+
+
+The following code sites use `Flush` today:
+- https://github.com/kubernetes/kubernetes/blob/ba57653b676fa639f86345f2bca20bc9e1bdc8a9/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go#L194
+- https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/util/flushwriter/writer.go#L50
+
+
+### Termination of the Request Handler
+
+Upon timeout, a request handler can self-terminate by returning the error it receives while:
```

Comment:
Unfortunately, when I experimented with this approach in the past, I discovered that the `context` package would call the `Value` method under certain circumstances from other go routines, so a `panic` in there would kill the whole process.  The way I mitigated that was checking that the current caller was in a http handler stack before `panic`-ing.  There might be other approaches as well.

Reply:
On he other hand, if we want to do post-timeout logging/reporting metric, we would need to access the request scoped attributes using the context, and the panic from `Value` is not desirable there.

Reply:
I think you could always special case whatever case you needed to do preserve.

Probably it would be best to experiment and measure how much benefit we see in terms of resource usage by pre-empting earlier via the `Value` method vs just the request writer.  Then you can determine if it is worth the effort to implement.